iquip/apps/dataviewer.py

```diff
@@ -820,69 +810,127 @@ class _DatasetFetcherThread(QThread):
         self.port = port
-        self.fetched.connect(callback, type=Qt.QueuedConnection)
+        self.mutex = QMutex()
+        self.modifyDone = QWaitCondition()
+        self._running = True
 
-    def run(self):
-        """Overridden."""
-        try:
-            response = requests.get(f"http://{self.ip}:{self.port}/dataset/master/",
-                                    params={"key": self.name},
-                                    timeout=10)
-            response.raise_for_status()
-        except requests.exceptions.RequestException:
-            logger.exception("Failed to fetch the dataset %s.", self.name)
-            return
-        dataset = np.array(response.json())
-        try:
-            response = requests.get(f"http://{self.ip}:{self.port}/dataset/master/",
-                                    params={"key": f"{self.name}.parameters"},
-                                    timeout=10)
-            response.raise_for_status()
-        except requests.exceptions.RequestException:
-            logger.exception("Failed to fetch the dataset parameters.")
-            parameters = list(map(str, range(dataset.shape[1] - 1)))
-        else:
-            parameters = response.json()
+    def _get(
+        self,
+        path: str,
+        params: Dict[str, Any],
+        default: Any = None,
+        timeout: float = 10
+    ) -> Any:
+        """Returns the json()-ed response of a GET request.
+        
+        Args:
+            path: The API path, i.e., the request url is "http://{ip}:{port}/{path}".
+            params: Params argument for the GET request.
+            default: The return value of this method when an exception occurs
+              during the GET request.
+            timeout: Timeout argument for the GET request.
+        """
+        url = f"http://{self.ip}:{self.port}/{path}"
         try:
-            response = requests.get(f"http://{self.ip}:{self.port}/dataset/master/",
-                                    params={"key": f"{self.name}.units"},
-                                    timeout=10)
+            response = requests.get(url, params=params, timeout=timeout)
             response.raise_for_status()
         except requests.exceptions.RequestException:
-            logger.exception("Failed to fetch the dataset units.")
-            units = [None] * (dataset.shape[1] - 1)
+            logger.exception("Failed to GET %s", url)
+            return default
+        return response.json()
+
+    def _initialize(self) -> float:
+        """Fetches the target dataset to initialize the local dataset.
+        
+        Returns:
+            The received timestamp or -1 if it failed.
+        """
+        response = self._get("dataset/master/", {"key": self.name})
+        if response is None or response[0] < 0:
+            return -1
+        timestamp, rawDataset = response
+        dataset = np.array(rawDataset)
+        numberOfParameters = dataset.shape[1] if dataset.ndim > 1 else 0
```

Comment:
It is just a question, but if `dataset.ndim = 0`, what happens in the data viewer?
I can't imagine this case🥲

Reply:
Also, which is better between `numParams` and the origianl one?

Reply:
When `ndim=0`, maybe the data policy can handle that case, but the viewer will not be able to.
It would crash I guess, and it is hard to define how it should react to such a case.
For now, I have no plans to support that case.

Reply:
> Also, which is better between `numParams` and the origianl one?

That's a good and popular abbreviation, but I used the full version since I already used `parameters`.